File: 2016/day_07.py
lines = open('./resources/input-7').readlines()

def split(line):
	brackets = []
	non_brackets = []
	adding_brackets = False
	while len(line) > 0:
		if adding_brackets:
			val, line = line.split(']', 1)
			brackets.append(val)
		else:
			val = line.split('[', 1)

			
			non_brackets.append(val[0])

			if len(val) == 1:
				break

			line = val[1]

		adding_brackets = not adding_brackets

	return (brackets, non_brackets)

def abba_detector(section):
	for i, _ in enumerate(section[0:-3]):
		a, b, c, d = section[i:i+4]

		if a != b and a == d and b == c:
			return True
	
	return False

def aba_detector(section):
	answers = []
	for i, _ in enumerate(section[0:-3]):
		a, b, c = section[i:i+3]

		if a != b and a == c:
			answers.append(section[i:i+3])
	
	return answers

def aba_to_bab(aba):
	a, b, _ = aba
	return "{}{}{}".format(b,a,b)

def supports_tls(line):
	brackets, non_brackets = split(line)
	brackets_okay = all(map(lambda x: not abba_detector(x), brackets))

	return  brackets_okay and any(map(lambda x: abba_detector(x), non_brackets))

def supports_ssl(line):
	brackets, non_brackets = split(line)
	aba = False

	for section in non_brackets:
		aba = aba_detector(section)
		if len(aba) > 0:
			# TODO: Aba is now an array of possible abas. Loop over and make this stuff complex
			bab = aba_to_bab(aba)
			if any(map(lambda x: bab in x, brackets)):
				return True

	return False
	# return (aba != False and any(map(lambda x: aba_to_bab(aba) in x, brackets)))
	# if aba == False:
	# 	return False


	# return  brackets_okay and any(map(lambda x: abba_detector(x), non_brackets))

i = 0
j = 0

for line in lines:
	if supports_tls(line):
		i = i + 1
	if supports_ssl(line):
		j = j + 1

print(i)
print(j)

File: 2016/day_07.rb
input = File.readlines("./resources/input-7").map(&:strip)

def supports_tls(txt)
  in_brackets = false
  tls = false
  txt.chars.each_cons(4) do |a,b,c,d|
    if d == "["
      in_brackets = true
      next
    elsif d == "]"
      in_brackets = false
      next
    end

    if a == d && b == c && a != b
      return false if in_brackets
      tls = true
    end
  end

  return tls
end


p input.count { |i| supports_tls i }

def supports_ssl(txt)
  in_brackets = false
  in_arr = []
  out_arr = []
  txt.chars.each_cons(3) do |a,b,c|
    if c == "["
      in_brackets = true
      next
    elsif c == "]"
      in_brackets = false
      next
    end

    if a == c && a != b
      if in_brackets
        in_arr << "#{b}#{a}"

        return true if out_arr.include? in_arr[-1]
      else
        out_arr << "#{a}#{b}"
        return true if in_arr.include? out_arr[-1]
      end
    end
  end

  return false
end

p input.count { |i| supports_ssl i }
